drivers/temperature/ltc2983/ltc2983.c

```diff
@@ -0,0 +1,644 @@
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <errno.h>
+#include "ltc2983.h"
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
+#include "no_os_print_log.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/******************************************************************************/
+
+/**
+ * @brief Device and comm init function
+ * @param device - LTC2983 descriptor to be initialized
+ * @param init_param - Init parameter for descriptor
+ * @return 0 in case of success, errno errors otherwise
+ */
+int ltc2983_init(struct ltc2983_desc **device,
+		 struct ltc2983_init_param *init_param)
+{
+	int ret, i;
+	struct ltc2983_desc *descriptor;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	ret = no_os_spi_init(&descriptor->comm_desc, &init_param->spi_init);
+	if (ret)
+		goto free_err;
+
+	if (init_param->dev_type == ID_LTC2983)
+		descriptor->max_channels_nr = 20;
+	else if (init_param->dev_type == ID_LTC2984)
+		descriptor->max_channels_nr = 20;
+	else if (init_param->dev_type == ID_LTC2986)
+		descriptor->max_channels_nr = 10;
+	descriptor->mux_delay_config_us = init_param->mux_delay_config_us;
+	descriptor->filter_notch_freq = init_param->filter_notch_freq;
+
+	descriptor->num_channels = 0;
+	for (i = 0; i < descriptor->max_channels_nr; i++) {
+		descriptor->sensors[i] = init_param->sensors[i];
+		if (init_param->sensors[i] &&
+		    init_param->sensors[i]->type != LTC2983_RSENSE)
+			descriptor->num_channels++;
+	}
+	if (!descriptor->num_channels) // should be at least one channel
+		goto spi_err;
+
+	ret = no_os_gpio_get(&descriptor->gpio_rstn, &init_param->gpio_rstn);
```

Comment:
Can the LTC2983 be software reset? In that case, we can make the reset GPIO optional